cmd/ops_agent_uap_plugin/service_linux.go

```diff
@@ -130,6 +157,103 @@ func (ps *OpsAgentPluginServer) GetStatus(ctx context.Context, msg *pb.GetStatus
+// This makes sure that GetStatus() returns a non-healthy status, signaling UAP to Start() the plugin again.
+//
+// ctx: the parent context that all child goroutines share.
+//
+// cancel: the cancel function for the parent context. By calling this function, the parent context is canceled,
+// and GetStatus() returns a non-healthy status, signaling UAP to re-trigger Start().
+func runSubagents(ctx context.Context, cancel context.CancelFunc, pluginInstallDirectory string, pluginStateDirectory string, restartCommand RestartCommandFunc, runCommand RunCommandFunc) {
+	// Register signal handler and implements its callback.
+	sigHandler(ctx, func(_ os.Signal) {
+		cancel()
+	})
+
+	var wg sync.WaitGroup
+	// Starting the diagnostics service
+	runDiagnosticsCmd := exec.CommandContext(ctx,
+		path.Join(pluginInstallDirectory, DiagnosticsBinary),
+		"-config", OpsAgentConfigLocationLinux,
+	)
+	wg.Add(1)
+	go restartCommand(ctx, cancel, runDiagnosticsCmd, runCommand, &wg)
+
+	// Starting Otel
+	runOtelCmd := exec.CommandContext(ctx,
+		path.Join(pluginInstallDirectory, OtelBinary),
+		"--config", path.Join(pluginStateDirectory, OtelRuntimeDirectory, "otel.yaml"),
+	)
+	wg.Add(1)
+	go restartCommand(ctx, cancel, runOtelCmd, runCommand, &wg)
+
+	// Starting FluentBit
+	runFluentBitCmd := exec.CommandContext(ctx,
+		path.Join(pluginInstallDirectory, AgentWrapperBinary),
+		"-config_path", OpsAgentConfigLocationLinux,
+		"-log_path", path.Join(pluginStateDirectory, LogsDirectory, "subagents/logging-module.log"),
+		path.Join(pluginInstallDirectory, FluentbitBinary),
+		"--config", path.Join(pluginStateDirectory, FluentBitRuntimeDirectory, "fluent_bit_main.conf"),
+		"--parser", path.Join(pluginStateDirectory, FluentBitRuntimeDirectory, "fluent_bit_parser.conf"),
+		"--storage_path", path.Join(pluginStateDirectory, FluentBitStateDiectory, "buffers"),
+	)
+	wg.Add(1)
+	go restartCommand(ctx, cancel, runFluentBitCmd, runCommand, &wg)
+
+	wg.Wait()
+}
+
+func restartCommand(ctx context.Context, cancel context.CancelFunc, cmd *exec.Cmd, runCommand RunCommandFunc, wg *sync.WaitGroup) {
+	defer wg.Done()
+	if cmd == nil {
+		return
+	}
+	if ctx.Err() != nil {
+		// context has been cancelled
+		log.Printf("cannot execute command: %s, because the context has been cancelled", cmd.Args)
+		return
+	}
+
+	childCtx, childCtxCancel := context.WithCancel(ctx)
+	defer childCtxCancel()
+
+	cmdCopy := exec.CommandContext(childCtx, cmd.Path, cmd.Args[1:]...)
+	cmdCopy.Env = cmd.Env
+	output, err := runCommand(cmdCopy)
+	if err != nil {
+		log.Printf("command: %s exited with errors, not restarting.\nCommand output: %s\n Command error:%s", cmd.Args, string(output), err)
+		cancel() // cancels the parent context which also stops other Ops Agent sub-binaries from running.
+		return
+	} else {
+		log.Printf("command: %s %s exited successfully.\nCommand output: %s", cmd.Path, cmd.Args, string(output))
+	}
+
+	wg.Add(1)
+	go restartCommand(ctx, cancel, cmd, runCommand, wg)
```

Comment:
Why do we make a recursive call here? What's the scenario where we need this?

This line will only be executed if `runCommand` exited without errors. The only scenario I can think of for that is when the subagents exit due to an exit signal.

Reply:
The command will be restarted when the previous execution exited without errors. 

I didn't think about when a command would exit without errors. This is just to catch that when it happens, and also trying to mimic the behaviour of `Restart=always` in systemd services.

Reply:
I think it's better if we don't restart if the subagent exited without errors, because the fact that the subagent exited (aside from during shutdown) is itself an error, so not restarting (resulting in giving an error to UAP in GetStatus) can help us catch it.

Also, since we are delegating the entire logic of crash recovery and restarting to UAP, I think we no longer need to mimic systemd's behaviour.